pyat/at/acceptance/momap_alternative.py

```diff
@@ -0,0 +1,402 @@
+"""momentum aperture alternative."""
+
+from __future__ import annotations
+
+import time
+from typing import Dict
+
+import numpy
+
+from ..lattice.lattice_object import Lattice
+
+__all__ = ["momaperture_project2start", "projectrefpts"]
+
+# 2024may29 oblanco at ALBA CELLS. First working version
+#           Based on the MATLAB implementation by Z.Marti at ALBA
+#           See https://github.com/atcollab/at/pull/773
+
+def momaperture_project2start(ring: Lattice, **kwargs: Dict[str, any]) -> numpy.ndarray:
+    """
+    :py:func:`momap_project2start` calculates the local momemtum aperture.
+
+    It is a binary search of the negative and positive energy thresholds
+    of stability around the closed orbit.
+
+    For a given energy offset the particles are first tracked from every
+    reference point to the end of the ring, and then, all or a set of particles
+    with different 6D coordinates are tracked together.  The surviving particles
+    continue the boundary search with a new energy step until the boundary is
+    found when the step limit or the  energy limit is reached.
+
+    Usage:
+      >>> momaperture_project2start(ring)
+
+    Parameters:
+      ring: list of elements
+
+    Keyword Arguments:
+      refpts: Selects the location of coordinates output.
+        See ":ref:`Selecting elements in a lattice <refpts>`"
+      nturns: number of turns to be tracked. Default 1000
+      dptol: energy offset resolution. Default 1e-4
+      euguess: maximum energy boundary. Default: the rf bucket height.
+      troffset: (2, N) offsets to be added to the transverse coordinates
+        on the N reference points. Default 1e-5 m
+      orbit: (N,6) offsets to be added on the N reference points.
+        Default, the closed orbit
+      verbose: print in the standard output additional info. Default False
+      epsilon6d: float.
+        If not passed, all particles are tracked.
+        If epsilon6d is given, we track for nturns only particles
+        having 6D coordinates different by epsilon6d
+
+    Returns:
+      dnp: (N,2) array with negative and positive stable energy boundaries
+        for the N reference points
+
+    ..note::
+      * This function could track in parallel. Set use_mp=True.
+        Other arguments could be passed, check :py:func:`lattice_track`.
+      * This function does a quick search, but, it is succeptible to miss
+        islands of instability due to the varying energy step.
+    """
+    # verboseprint to check flag only once
+    verbose = kwargs.pop("verbose", False)
+    verboseprint = print if verbose else lambda *a, **k: None
+
+    rps = kwargs.pop("refpts", ring.uint32_refpts(range(len(ring))))
+    nrps = len(rps)
```

Comment:
`refpts` may be a string (element name pattern), or an element class, or a boolean mask. Before considering it as a list of locations, you must convert it with:
```python
rps = ring.get_uint32_index(rps)
```

Reply:
@lfarv , thank you. It has been modified in the last commit
```bash
     rps = kwargs.pop("refpts", ring.uint32_refpts(range(len(ring))))
+    rps = ring.get_uint32_index(rps)
     nrps = len(rps)
```